design/task-management-api.md

```diff
@@ -0,0 +1,346 @@
+    +----------+
+         |
+         |
+    +----v-----+
+    |  task 2  |
+    +----+-----+
+         |
+         |
+    +----v-----+
+    |  task 3  |
+    +----------+
+```
+
+##### Chords (and Groups)
+
+```
+                    +--------+
+               +--->|  task1 +-----+
+               |    +--------+     |
+               |                   |
+               |                   |
+               |    +--------+     |
+               +--->|  task2 +-----+
+               |    +--------+     |
+ +-------+     |                   |     +-------------+
+ |  prev |-----+                   +---->|(spontaneous)|
+ +-------+     |    +--------+     |     +-------------+
+               +--->|  task3 +-----+
+               |    +--------+     |
+               |                   |
+               |                   |
+               |    +--------+     |
+               +--->|  task4 +-----+
+                    +--------+
+```
+
+## Implementing "user" stories with the API
+
+### Branch export
+
+1. Under the merge/commit lock for the branch:
+  1. Insert a task with ID `start-export-{branch}` to start generating export tasks and a task
+     `done-export-{branch}` that depends on it.
+  1. If insertion failed, _replace_ the task with ID `next-export-{branch}` with a task to
+     export _this_ comit ID.  And add a dependency on `done-export-{branch}` (which may fail if
+     that task has completed).
+1. To handle `start-export-{branch}`:
+   1. Generate a task to copy or delete each file object (this is an opportunity to batch
+      multiple file objects if performance doesn't match.  `done-export-{branch}` depends on
+      each of these tasks (and cannot have been deleted since `start-export-{branch}` has not
+      yet been returned).  For every prefix for which such an object is configured, add a task
+      to generate its `.../_SUCCESS` object on S3, dependent on all the objects under that
+      prefix (or, to handle objects in sub-prefixes, on just the `_SUCCESS` of that sub-prefix).
+   2. Add a task to generate manifests, dependent on `done-export-{branch}`.
+   2. Return `start-export-{branch}` as completed.
+1. To handle a copy or delete operation, perform it.
+1. To handle `done-export-{branch}`: just return it, it can be spontaneous (if the task queue
+   supports that).
+1. To handle `next-export-{branch}`: create `start-export-{branch}` (the previous one must have
+   ended), and return the task.
+
+`next-export-{branch}` is used to serialize branch exports, achieving the requirement for single
+concurrent export per branch.  Per-prefix `_SUCCESS` objects are generated on time due to their
+dependencies.  (As an option, we could set priorities and return tasks in priority order from
+`OwnTasks`, to allow `_SUCCESS` objects to be created before copying other objects.)  Retries
+are handled by setting multiple per-copy attempts.
+
+### Concurrent dedupe batching
+
+1. For every created object, create a dedupe task.
```

Comment:
I think I'm missing something about dedupe.  I thought dedupe was after CreateObject returned because of processing rate, and the DB was _not_ the limiting factor?
Specifically, If we only returned from CreateEntry once dedupe were finished, we would have a single write to the entries table and a single write to objects_dedupe on CreateEntry; we could batch those just like we batch the post-dedupe writes.  Instead, we write once to entries to CreateEntry, then again to objects_dedupe and to entries once dedupe is finished.  That's 3 operations (2 batched) rather than 2 (that could be batched in the same way) -- and we measured a speedup!  How can the DB be blocking us here?

Reply:
Dedup is pure DB - no processing.
It was moved from the entry writing transaction to speed-up response to the user, not to improve throughput.
the third action (update to the entry) is done only when a duplicate is found. I expect this not to be a large part of processing. and anyway - it does not slow response to user.

Reply:
Let's remove it from here, then, at least for now.